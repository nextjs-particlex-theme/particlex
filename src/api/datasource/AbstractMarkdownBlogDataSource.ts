import path from 'node:path'
import {
  generateShallowToc,
  highlight,
  markdownToHtml, replacePrefixIndent,
  utilMarkdownGenerateCaster
} from '@/api/datasource/util'
import readline from 'node:readline'
import fs from 'node:fs'
import yaml, { YAMLParseError } from 'yaml'
import { globSync } from 'glob'
import type { Category, DataSourceConfig, Tag } from '@/api/datasource/types/definitions'
import { StaticResource } from '@/api/datasource/types/resource/StaticResource'
import type { SEO } from '@/api/datasource/types/resource/Post'
import Post from '@/api/datasource/types/resource/Post'
import type { BlogDataSource } from '@/api/datasource/types/BlogDataSource'
import cached from '@/lib/cached'

type AbstractBlogDataSourceCons = {
  /**
   * 首页文章目录.
   */
  homePostDirectory: string
  /**
   * 所有文章的共同根目录
   */
  postDirectory: string
  /**
   * 资源目录
   */
  resourceDirectory: string
}

type AbstractBlogDatasourceConfig = {
  /**
   * 首页文章目录.
   */
  homePostGlob: string
  /**
   * 所有文章的共同根目录
   */
  postGlob: string
  /**
   * 资源目录
   */
  resourceGlob: string
}

type PostContent = {
  metadata: any
  content: string
}

type PostConstructor = ConstructorParameters<typeof Post>[0]

export default abstract class AbstractMarkdownBlogDataSource implements BlogDataSource {

  protected readonly config: AbstractBlogDatasourceConfig


  protected constructor(config: AbstractBlogDataSourceCons) {
    this.config = {
      homePostGlob: `${config.homePostDirectory}/**/*.md`,
      postGlob: `${config.postDirectory}/**/*.md`,
      resourceGlob: `${config.resourceDirectory}/**/*`
    }
  }


  protected buildCacheKey(path: string[]): string {
    return path.join('/')
  }


  /**
   * 获取配置
   */
  abstract getConfig(): Promise<DataSourceConfig>

  /**
   * 获取博客的 web 访问路径.
   * @param path 博客文件路径.
   */
  abstract resolvePostWebPath(path: string): string[]

  /**
   * 获取静态文件的 web 访问路径
   * @param path 静态文件路径. 需要注意操作系统的分隔符.
   */
  abstract resolveStaticResourceWebPath(path: string): string[]

  @cached()
  async getTagMapping(): Promise<Map<Tag, Post>> {
    const posts = await this.getAllPost()
    const r = new Map<Category, Post>()
    for (let post of posts) {
      for (let tag of post.tags) {
        r.set(tag, post)
      }
    }
    return r
  }

  @cached()
  async getCategoriesMapping(): Promise<Map<Category, Post>> {
    const posts = await this.getAllPost()
    const r = new Map<Category, Post>()
    for (let post of posts) {
      for (let category of post.categories) {
        r.set(category, post)
      }
    }
    return r
  }

  async pageHomePosts(page: number = 0, size: number = 5): Promise<Post[]> {
    const files = globSync(this.config.homePostGlob, { cwd: process.env.BLOG_PATH })
    return (await this.parseMarkdownFiles(files, page * size, size)).sort((a, b) => {
      const v1 = a.date ?? 0
      const v2 = b.date ?? 0
      return v2 - v1
    })
  }

  @cached()
  pagePostsSize(): Promise<number> {
    return Promise.resolve(globSync(this.config.homePostGlob, { cwd: process.env.BLOG_PATH }).length)
  }

  @cached()
  async getAllPost(): Promise<Post[]> {
    const files = globSync(this.config.postGlob, { cwd: process.env.BLOG_PATH })
    return await this.parseMarkdownFiles(files)
  }

  @cached()
  getAllStaticResource(): Promise<StaticResource[]> {
    const files = globSync(this.config.resourceGlob, { cwd: process.env.BLOG_PATH })
    return Promise.resolve(
      files.map(v =>
        new StaticResource(path.resolve(process.env.BLOG_PATH, v), this.resolveStaticResourceWebPath(v)))
    )
  }

  private parsePostContent(postPath: string): Promise<PostContent> {
    return new Promise((resolve, reject) => {
      const rl = readline.createInterface({
        input: fs.createReadStream(postPath),
      })
      const metadataStrArr: string[] = []
      // 0: expect start.
      // 1: expect end.
      // 2: collected.
      let metadataCollectStatus = 0
      const content: string[] = []

      rl.on('line', (line) => {
        if (metadataCollectStatus < 2) {
          metadataStrArr.push(line)
          if (line.startsWith('---')) {
            metadataCollectStatus++
          }
        } else {
          content.push(line)
        }
      })

      rl.on('close', () => {
        let metadata: any
        let html: string
        if (metadataCollectStatus < 2) {
          metadata = {}
          html = markdownToHtml(metadataStrArr.join('\n'))
        } else {
          try {
            metadata = yaml.parse(metadataStrArr.slice(1, metadataStrArr.length - 1).join('\n'))
          } catch (e) {
            if (e instanceof YAMLParseError && e.code === 'TAB_AS_INDENT') {
              const str = metadataStrArr.map(replacePrefixIndent).slice(1, metadataStrArr.length - 1).join('\n')
              try {
                metadata = yaml.parse(str)
              } catch (e) {
                if (e instanceof YAMLParseError && e.code === 'TAB_AS_INDENT') {
                  reject(new Error(`Failed to parse markdown file ${postPath}, 
                  neither modify environment variable \`YAML_INDENT_SPACE_COUNT\` or remove \`\\t\` in your yaml config.`, { cause: e }))
                  return
                }
              }
            } else {
              // man! what else can I do?
              reject(e)
              return
            }
          }
          html = markdownToHtml(content.join('\n'))
        }
        resolve({
          metadata: metadata ?? {},
          content: html
        })
      })
    })
  }

  /**
   * 解析 markdown 文件，返回解析后的内容
   * @param files 需要解析的文件
   * @param begin 从哪里开始
   * @param take 解析几个
   * @protected
   */
  protected async parseMarkdownFiles(files: string[], begin = 0, take?: number): Promise<Post[]> {
    const r: Post[] = []

    const len = take === undefined ? files.length : Math.min(take + begin, files.length)

    for (let i = Math.max(0, begin); i < len; i++) {
      const file = files[i]
      const { metadata, content } = await this.parsePostContent(path.resolve(process.env.BLOG_PATH, file))
      let source = this.resolvePostWebPath(file)

      const postData: PostConstructor = {
        title: metadata.title,
        date: metadata.date ? new Date(metadata.date).valueOf() : undefined,
        source,
        id: source.join('-'),
        toc: generateShallowToc(content, utilMarkdownGenerateCaster),
        content: highlight(content),
        tags: this.parseTagAndCategories(metadata.tags),
        categories: this.parseTagAndCategories(metadata.categories),
        // will init later before return.
        seo: (null as any)
      }
      postData.seo = this.parseSeoConfig(postData, metadata.seo)

      r.push(new Post(postData))
    }
    return r
  }

  private parseTagAndCategories(val?: any): string[] {
    if (!val) {
      return []
    }
    if (Array.isArray(val)) {
      return val
    }
    if (typeof val === 'string') {
      return [val]
    }
    // 兼容历史遗留写法. 这个写法是之前 hexo particlex 主题特有的
    //  categories:
    //    data:
    //      - { name: "Hello", path: "/hello/" }
    if (!val.data || !Array.isArray(val.data)) {
      return []
    }
    const r: string[] = []
    for (let datum of val.data) {
      if (typeof datum.name === 'string') {
        r.push(datum.name)
      }
    }
    return r
  }

  /**
   * 解析 SEO 配置
   * @param data 已有的构造器参数
   * @param seo seo 配置
   * @protected
   */
  protected parseSeoConfig(data: Omit<PostConstructor, 'seo'>, seo?: Record<string, undefined | string>): SEO {
    const fakeType = (seo ?? {}) as Partial<SEO>

    return {
      title: fakeType.title ?? data.title,
      description: fakeType.description,
      // TODO auto generate tags.
      keywords: fakeType.keywords ?? []
    }
  }

  @cached()
  private async postMap(): Promise<Map<string, Post>> {
    const r = await this.getAllPost()
    const result = new Map<string, Post>()
    for (let p of r) {
      result.set(this.buildCacheKey(p.source), p)
    }
    return result
  }
  
  async getPostByWebUrl(url: string[]): Promise<Post | undefined> {
    return (await this.postMap()).get(this.buildCacheKey(url))
  }

  @cached()
  private async staticResourceMap(): Promise<Map<string, StaticResource>> {
    const r = await this.getAllStaticResource()
    const resourceCache = new Map<string, StaticResource>()
    for (let staticResource of r) {
      resourceCache.set(this.buildCacheKey(staticResource.accessPath), staticResource)
    }
    return resourceCache
  }

  async getStaticResourceByWebUrl(url: string[]): Promise<StaticResource | undefined> {
    return (await this.staticResourceMap()).get(this.buildCacheKey(url))
  }
}