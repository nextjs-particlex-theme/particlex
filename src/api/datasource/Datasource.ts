import type { DataSourceConfig } from '@/api/datasource/types/definitions'

export type DatasourceItem = {
  /**
   * 唯一标识符, 通常是文件访问路径
   */
  id: string;
  /**
   * 访问路径
   */
  visitPath: WebVisitPath
  /**
   * 文件类型，可以是文件路径，或者文件类型拓展符
   */
  type: string
}

export type WebVisitPath = string[]

export type StaticResourceContent = {
  base64: string
  contentType: string
}

export interface Datasource {
  /**
   * 获取配置
   */
  getConfig(): Promise<Readonly<DataSourceConfig>>

  /**
   * 获取所有的文章
   * @return {string} 文章唯一标识符
   */
  getAllPages(): Promise<Array<DatasourceItem>>

  /**
   * 获取页面内容
   */
  getPage(id: DatasourceItem['id']): Promise<string>

  /**
   * 获取首页文章
   */
  getAllHomePosts(): Promise<Array<DatasourceItem>>

  /**
   * 获取所有静态资源
   */
  getAllStaticResource(): Promise<Array<DatasourceItem>>

  /**
   * 获取资源的内容
   * @return base64 字符串
   */
  getResource(id: DatasourceItem['id']): Promise<StaticResourceContent | undefined>
}