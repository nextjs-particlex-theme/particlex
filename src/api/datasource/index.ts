import hexo from './hexo/index'
import type { BlogDataSource } from '@/api/datasource/types/BlogDataSource'

let datasource: BlogDataSource

if (process.env.DATASOURCE_CACHE_ENABLE === 'true') {
  const cacheHolder: Record<string, any> = {}

  // eslint-disable-next-line no-undef
  datasource = new Proxy(hexo, {
    get(target: BlogDataSource, p: string | symbol): any {
      // @ts-ignore
      const orig = target[p]
      if (typeof orig === 'function') {
        return function (...args: unknown[]) {
          const key = orig.name + ':' + args.join('_')
          if (cacheHolder[key]) {
            return cacheHolder[key]
          } else {
            const r = orig.apply(target, args)
            if (r.then && r.catch) {
              return r.then((result: unknown) => {
                cacheHolder[key] = result
                return result
              })
            }
            return cacheHolder[key] = orig.apply(target, args)
          }
        }
      } else {
        return orig
      }
    }
  })
} else {
  datasource = hexo
}


export default datasource
