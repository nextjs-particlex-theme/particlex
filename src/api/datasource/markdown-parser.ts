import readline from 'node:readline'
import fs from 'node:fs'
import yaml, { YAMLParseError } from 'yaml'
import showdown from 'showdown'
import type { TocItem } from '@/api/datasource/types/definitions'
import { JSDOM } from 'jsdom'


showdown.setFlavor('github')

export type PostContent = {
  metadata: {
    title?: string
    date?: string
    categories?: string | string[]
    tags?: string | string[]
    seo?: {
      title?: string
      description?: string
      keywords?: string | string[]
    }
  }
  content: string
  toc: TocItem[]
}

const LEVEL_MAPPING: Record<string, number> = {
  H1: 0,
  H2: 1,
  H3: 2,
  H4: 3,
  H5: 4,
  H6: 5,
}


/**
 * 根据 html 自动生成 Toc，要求必须平铺 h1 h2 等标签。
 * @see [/__tests__/api/util.test.ts](/__tests__/api/util.test.ts)
 */
function generateShallowToc(html?: string): TocItem[] {
  if (!html) {
    return []
  }
  const dom = new JSDOM(html, { contentType: 'text/html' })

  const root = dom.window.document.body

  let parentStack: TocItem[] = [{ title: 'FakeRoot', child: [], anchor: '#' }]
  root.childNodes.forEach(v => {
    const currentLevel = LEVEL_MAPPING[v.nodeName]
    if (currentLevel === undefined) {
      return
    }
    const heading = v as HTMLHeadingElement
    if (parentStack[currentLevel]) {
      // 父节点存在
      // 弹出多余节点
      for (let i = currentLevel + 1; i < parentStack.length; i++) {
        parentStack.pop()
      }
      let item: TocItem
      const data = {
        title: heading.innerHTML,
        anchor: '#' + heading.getAttribute('id')
      }
      item = {
        ...data,
        child: []
      }

      parentStack[currentLevel].child.push(item)
      parentStack[currentLevel + 1] = item
    }
  })
  return parentStack[0].child
}

let __test_generateShallowToc0: typeof generateShallowToc | undefined
if (process.env.NODE_ENV === 'test') {
  __test_generateShallowToc0 = generateShallowToc
} else {
  __test_generateShallowToc0 = undefined
}

/**
 * Test only
 */
export const __test_generateShallowToc = __test_generateShallowToc0




const YAML_INDEX_SPACES_FILL: string = (new Array(Number.parseInt(process.env.YAML_INDENT_SPACE_COUNT, 10))).map(() => ' ').join('')

/**
 * 将 `\t` 替换为 `两个空格`.<p>
 * 用于处理 yaml 中使用了 `\t` 作为缩进的问题
 * @param str
 */
const replacePrefixIndent = (str: string): string => {
  let i = 0
  while (i < str.length && str.charAt(i) === '\t') {
    i++
  }
  if (i === str.length) {
    return ''
  }
  if (i === 0) {
    return str
  }
  const spaces: string[] = []
  for (let j = 0; j < i; j++) {
    spaces.push(YAML_INDEX_SPACES_FILL)
  }
  return spaces.join('') + str.substring(i)
}

const REPLACE_REGX = /^\t+/g

/**
 * 创建一个错误，指出哪里使用了 TAB
 */
function indicateWhereContainsTab(yamlLines: string[], filePath: string): Error {
  let msg: string[] = [`Failed to parse YAML in the markdown file '${filePath}'. The lines below start with a TAB (TABs are replaced by a '→'):`]
  for (let i = 0; i < yamlLines.length; i++) {
    const line = yamlLines[i]
    if (line.startsWith('\t')) {
      msg.push(`\tLine ${i + 1}, content: ${line.replaceAll(REPLACE_REGX, '→')}`)
    }
  }
  msg.push('\nYou can either replace them with spaces or modify the environment variable \'YAML_INDENT_SPACE_COUNT\' to fix it automatically.')
  return new Error(msg.join('\n'))
}

/**
 * markdown 转 html
 */
const markdownToHtml = (markdownContent: string): string => {
  const sd = new showdown.Converter({
    strikethrough: true,
    tables: true,
    tasklists: true,
    disableForced4SpacesIndentedSublists: true,
    // rawPrefixHeaderId: true,
    rawHeaderId: true
  })
  return sd.makeHtml(markdownContent)
}

/**
 * 解析 markdown 文件
 * @param filepath 文件路径
 */
export const parseMarkdownFile = (filepath: string): Promise<PostContent> => {
  return new Promise((resolve, reject) => {
    const rl = readline.createInterface({
      input: fs.createReadStream(filepath),
    })
    const metadataStrArr: string[] = []
    // 0: expect start.
    // 1: expect end.
    // 2: collected.
    let metadataCollectStatus = 0
    const content: string[] = []

    rl.on('line', (line) => {
      if (metadataCollectStatus < 2) {
        metadataStrArr.push(line)
        if (line.startsWith('---')) {
          metadataCollectStatus++
        }
      } else {
        content.push(line)
      }
    })

    rl.on('close', () => {
      let metadata: any
      let html: string
      if (metadataCollectStatus < 2) {
        metadata = {}
        html = markdownToHtml(metadataStrArr.join('\n'))
      } else {
        const original = metadataStrArr.slice(1, metadataStrArr.length - 1)
        try {
          metadata = yaml.parse(original.join('\n'))
        } catch (e) {
          if (e instanceof YAMLParseError && e.code === 'TAB_AS_INDENT') {
            const str = metadataStrArr.map(replacePrefixIndent).slice(1, metadataStrArr.length - 1).join('\n')
            try {
              metadata = yaml.parse(str)
            } catch (e) {
              reject(indicateWhereContainsTab(original, filepath))
              return
            }
          } else {
            reject(new Error(`Parse yaml file '${filepath}' failed, content:\n${original.join('\n')}`, { cause: e }))
            return
          }
        }
        html = markdownToHtml(content.join('\n'))
      }
      resolve({
        metadata: metadata ?? {},
        content: html,
        toc: generateShallowToc(html)
      })
    })
  })
}