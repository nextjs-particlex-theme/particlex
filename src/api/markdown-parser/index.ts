import mdParser from '@/api/markdown-parser/impl/md/MdParser'
import mdxParser from '@/api/markdown-parser/impl/mdx/MdxParser'
import type { ParsedMarkdown } from '@/api/markdown-parser/types'
import os from 'node:os'
import yaml, { YAMLParseError } from 'yaml'

/**
 * 解析markdown文本内容
 * @param markdown markdown 文本内容
 * @param identifier 标识符，md或mdx
 */
const parseMarkdown = (markdown: string, identifier: string): Promise<ParsedMarkdown> => {
  if (identifier.endsWith('.md')) {
    return mdParser.parse(markdown)
  } else if (identifier.endsWith('.mdx')) {
    // TODO error handle.
    return mdxParser.parse(markdown)
  } else {
    throw new Error('Unknown file.')
  }
}

export type Markdown = {
  metadata: {
    title?: string
    date?: string
    categories?: string | string[]
    tags?: string | string[]
    seo?: {
      title?: string
      description?: string
      keywords?: string | string[]
    }
  }
  content: string
}

const YAML_INDEX_SPACES_FILL: string = (new Array(Number.parseInt(process.env.YAML_INDENT_SPACE_COUNT, 10))).map(() => ' ').join('')

/**
 * 将 `\t` 替换为 `两个空格`.<p>
 * 用于处理 yaml 中使用了 `\t` 作为缩进的问题
 * @param str
 */
const replacePrefixIndent = (str: string): string => {
  let i = 0
  while (i < str.length && str.charAt(i) === '\t') {
    i++
  }
  if (i === str.length) {
    return ''
  }
  if (i === 0) {
    return str
  }
  const spaces: string[] = []
  for (let j = 0; j < i; j++) {
    spaces.push(YAML_INDEX_SPACES_FILL)
  }
  return spaces.join('') + str.substring(i)
}



enum CollectStatus {
  EXPECT_START,
  EXPECT_END,
  DONE
}

const REPLACE_REGX = /^\t+/g


/**
 * 创建一个错误，指出哪里使用了 TAB
 */
function indicateWhereContainsTab(yamlLines: string[], filePath: string): Error {
  let msg: string[] = [`Failed to parse YAML in the markdown file '${filePath}'. The lines below start with a TAB (TABs are replaced by a '→'):`]
  for (let i = 0; i < yamlLines.length; i++) {
    const line = yamlLines[i]
    if (line.startsWith('\t')) {
      msg.push(`\tLine ${i + 1}, content: ${line.replaceAll(REPLACE_REGX, '→')}`)
    }
  }
  msg.push('\nYou can either replace them with spaces or modify the environment variable \'YAML_INDENT_SPACE_COUNT\' to fix it automatically.')
  return new Error(msg.join('\n'))
}


/**
 * 解析 Markdown 文本内容
 * @param content Markdown内容，提供一个以换行符分割的数组或者整个字符串，后者将会被转化为前者
 * @param filepath 文件路径，当解析 markdown 错误时，将会带上文件路径以便于排查
 */
export const splitMarkdownContent = (content: string[] | string, filepath: string = '<Unknown>'): Markdown => {
  const metadataStrArr: string[] = []
  let metadataCollectStatus = CollectStatus.EXPECT_START
  let contentArr: string[] = Array.isArray(content) ? content : content.split(os.EOL)

  for (let line of contentArr) {
    if (metadataCollectStatus == CollectStatus.DONE) {
      break
    }
    switch (metadataCollectStatus) {
    case CollectStatus.EXPECT_START: {
      if (line.startsWith('---')) {
        metadataCollectStatus = CollectStatus.EXPECT_END
      }
      break
    }
    case CollectStatus.EXPECT_END: {
      if (line.startsWith('---')) {
        metadataCollectStatus = CollectStatus.DONE
        break
      }
      metadataStrArr.push(line)
      break
    }
    // eslint-disable-next-line no-fallthrough
    default:
      throw new Error('Unreachable branch!')
    }
  }


  if (metadataCollectStatus !== CollectStatus.DONE) {
    return {
      content: Array.isArray(content) ? content.join(os.EOL) : content,
      metadata: {}
    }
  }

  let metadata: Markdown['metadata']
  try {
    metadata = yaml.parse(metadataStrArr.join('\n'))
  } catch (e) {
    if (e instanceof YAMLParseError && e.code === 'TAB_AS_INDENT') {
      const str = metadataStrArr.map(replacePrefixIndent).join('\n')
      try {
        metadata = yaml.parse(str)
      } catch (e) {
        throw indicateWhereContainsTab(metadataStrArr, filepath)
      }
    } else {
      throw new Error(`Parse yaml file '${filepath}' failed, content:\n${metadataStrArr.join(os.EOL)}`, { cause: e })
    }
  }
  return {
    content: contentArr.slice(metadataStrArr.length + 2).join(os.EOL),
    metadata,
  }

}


export default parseMarkdown
