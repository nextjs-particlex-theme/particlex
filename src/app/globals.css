@tailwind base;
@tailwind components;
@tailwind utilities;

@layer base {
    :root {
        --color-white: #fff;
        --color-text-sub: #5c6b72;
        --color-text: #1e3e3f;
        --color-primary: #66afef;
        --color-error: #ef4444;
        --border-color: #D9D9D9FF;
        --z-index-top: 100000;
        --z-index-common: 10000;
        --background-color: #f6f8fa;
    }
}

body {
    background: var(--background-color);
    color: #1e3e3f;
    font: 500 14px Lexend, "Noto Sans SC", sans-serif;
    overflow-x: hidden;
    width: 100%;
}

.link-styled-container a {
    color: var(--color-primary);
    text-decoration: none;
}
a:hover {
    opacity: .8;
}

::-webkit-scrollbar {
    height: 12px;
    width: 12px;
}

::-webkit-scrollbar-thumb {
    background: #8ab5ff linear-gradient(
            45deg,
            #fff6 25%,
            transparent 25%,
            transparent 50%,
            #fff6 50%,
            #fff6 75%,
            transparent 75%,
            transparent
    );
    border: 3px solid #e6efff;
    border-radius: 100px;
}
::-webkit-scrollbar-track {
    background: #e6efff;
    border-radius: 100px;
}
::selection {
    background-color: #3392ff2a;
    color: unset;
}

@keyframes fadeIn {
    from {
        opacity: 0;
        transform: translateY(-20px) scale(1.1);
    }

    to {
        opacity: 1;
        transform: none;
    }
}

/*添加动画*/
#fade-in {
    animation-duration: .8s;
    animation-fill-mode: both;
    animation-name: fadeIn;
}