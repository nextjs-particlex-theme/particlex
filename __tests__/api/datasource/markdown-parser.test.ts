import { expect, test } from '@jest/globals'
import markdown from './markdown.test.md'
import badHeadingMarkdown from './bad-heading.test.md'
import { parseMarkdownContent } from '@/api/datasource/markdown-parser'

test('Test markdown parse', () => {
  const post = parseMarkdownContent(markdown)
  expect(post).toMatchSnapshot()
})


test('Test bad heading', () => {
  const post = parseMarkdownContent(badHeadingMarkdown)
  expect(post).toMatchSnapshot()
})