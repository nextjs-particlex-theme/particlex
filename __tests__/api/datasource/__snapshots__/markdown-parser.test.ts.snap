// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test bad heading 1`] = `
{
  "content": "<h4 id="l4">L4</h4>
<h3 id="l3">L3</h3>
<h2 id="l2">L2</h2>
<h1 id="l1">L1</h1>
<h2 id="l2-1">L2</h2>
<h3 id="l3-1">L3</h3>
<h4 id="l4-1">L4</h4>",
  "metadata": {},
  "toc": [
    {
      "anchor": "#l4",
      "child": [],
      "title": "L4",
    },
    {
      "anchor": "#l3",
      "child": [],
      "title": "L3",
    },
    {
      "anchor": "#l2",
      "child": [],
      "title": "L2",
    },
    {
      "anchor": "#l1",
      "child": [
        {
          "anchor": "#l2-1",
          "child": [
            {
              "anchor": "#l3-1",
              "child": [
                {
                  "anchor": "#l4-1",
                  "child": [],
                  "title": "L4",
                },
              ],
              "title": "L3",
            },
          ],
          "title": "L2",
        },
      ],
      "title": "L1",
    },
  ],
}
`;

exports[`Test markdown parse 1`] = `
{
  "content": "<p>Markdown是一种轻量级的「标记语言」。</p>
<h2 id="1-markdown语法教程">1 Markdown语法教程</h2>
<h3 id="11-标题">1.1 标题</h3>
<p>不同数量的<code>#</code>可以完成不同的标题.。</p>
<h3 id="12-字体">1.2 字体</h3>
<p>粗体、斜体、粗体和斜体，删除线，需要在文字前后加不同的标记符号。如下：</p>
<p><strong>这个是粗体</strong></p>
<p><em>这个是斜体</em></p>
<p><strong><em>这个是粗体加斜体</em></strong></p>
<p><del>这里是删除线</del></p>
<p>注：如果想给字体换颜色、字体或者居中显示，需要使用内嵌HTML来实现。</p>
<h3 id="13-无序列表">1.3 无序列表</h3>
<p>无序列表的使用，在符号<code>-</code>后加空格使用。如下：</p>
<ul>
<li>无序列表 1</li>
<li>无序列表 2</li>
<li>无序列表 3</li>
</ul>
<p>如果要控制列表的层级，则需要在符号<code>-</code>前使用空格。如下：</p>
<ul>
<li>无序列表 1</li>
<li>无序列表 2<ul>
<li>无序列表 2.1</li>
<li>无序列表 2.2</li></ul></li>
</ul>
<h3 id="14-有序列表">1.4 有序列表</h3>
<p>有序列表的使用，在数字及符号<code>.</code>后加空格后输入内容，如下：</p>
<ol>
<li>有序列表 1</li>
<li>有序列表 2</li>
<li>有序列表 3</li>
</ol>
<h3 id="15-引用">1.5 引用</h3>
<p>引用的格式是在符号<code>&gt;</code>后面书写文字。如下：</p>
<blockquote>
  <p>读一本好书，就是在和高尚的人谈话。 ——歌德</p>
</blockquote>
<blockquote>
  <p>雇用制度对工人不利，但工人根本无力摆脱这个制度。 ——阮一峰</p>
</blockquote>
<h3 id="16-链接">1.6 链接</h3>
<p><a href="https://github.com/nextjs-particlex-theme/particlex">particlex</a></p>
<h3 id="17-图片">1.7 图片</h3>
<p><img src="/favicon.ico" alt="favicon.ico" /></p>
<h3 id="18-分割线">1.8 分割线</h3>
<p>可以在一行中用三个以上的减号来建立一个分隔线，同时需要在分隔线的上面空一行。如下：</p>
<hr />
<h3 id="19-表格">1.9 表格</h3>
<p>可以使用冒号来定义表格的对齐方式，如下：</p>
<table>
<thead>
<tr>
<th id="姓名" style="text-align:left;">姓名</th>
<th id="年龄" style="text-align:center;">年龄</th>
<th id="工作" style="text-align:right;">工作</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">小可爱</td>
<td style="text-align:center;">18</td>
<td style="text-align:right;">吃可爱多</td>
</tr>
<tr>
<td style="text-align:left;">小小勇敢</td>
<td style="text-align:center;">20</td>
<td style="text-align:right;">爬棵勇敢树</td>
</tr>
<tr>
<td style="text-align:left;">小小小机智</td>
<td style="text-align:center;">22</td>
<td style="text-align:right;">看一本机智书</td>
</tr>
</tbody>
</table>
<h3 id="110-代码块">1.10 代码块</h3>
<p>如果在一个行内需要引用代码，只要用反引号引起来就好，如下：</p>
<p>Use the <code>printf()</code> function.</p>
<p>在需要高亮的代码块的前一行及后一行使用三个反引号，同时<strong>第一行反引号后面表示代码块所使用的语言</strong>，如下：</p>
<pre><code class="java language-java">// FileName: HelloWorld.java
public class HelloWorld {
  // Java 入口程序，程序从此入口
  public static void main(String[] args) {
    System.out.println("Hello,World!"); // 向控制台打印一条语句
  }
}</code></pre>
<pre><code class="xml language-xml">&lt;hello&gt;
    &lt;world&gt;&lt;/world&gt;
&lt;/hello&gt;</code></pre>
<h2 id="2-其他语法">2 其他语法</h2>
<h3 id="21-html">2.1 HTML</h3>
<p>支持原生 HTML 语法，请写内联样式，如下：</p>
<p><span style="display:block;text-align:right;color:orangered;">橙色居右</span><br />
<span style="display:block;text-align:center;color:orangered;">橙色居中</span></p>
<h3 id="github-警告">Github 警告</h3>
<blockquote>
  <p>[!NOTE]<br />
  Note 123</p>
</blockquote>
<blockquote>
  <p>[!TIP]<br />
  TIP 123</p>
</blockquote>
<blockquote>
  <p>[!IMPORTANT]<br />
  IMPORTANT 123</p>
</blockquote>
<blockquote>
  <p>[!WARNING]<br />
  WARNING 123</p>
</blockquote>
<blockquote>
  <p>[!CAUTION]<br />
  CAUTION 123</p>
</blockquote>",
  "metadata": {
    "date": "2024/07/31 16:22:39",
    "title": "example",
  },
  "toc": [
    {
      "anchor": "#1-markdown语法教程",
      "child": [
        {
          "anchor": "#11-标题",
          "child": [],
          "title": "1.1 标题",
        },
        {
          "anchor": "#12-字体",
          "child": [],
          "title": "1.2 字体",
        },
        {
          "anchor": "#13-无序列表",
          "child": [],
          "title": "1.3 无序列表",
        },
        {
          "anchor": "#14-有序列表",
          "child": [],
          "title": "1.4 有序列表",
        },
        {
          "anchor": "#15-引用",
          "child": [],
          "title": "1.5 引用",
        },
        {
          "anchor": "#16-链接",
          "child": [],
          "title": "1.6 链接",
        },
        {
          "anchor": "#17-图片",
          "child": [],
          "title": "1.7 图片",
        },
        {
          "anchor": "#18-分割线",
          "child": [],
          "title": "1.8 分割线",
        },
        {
          "anchor": "#19-表格",
          "child": [],
          "title": "1.9 表格",
        },
        {
          "anchor": "#110-代码块",
          "child": [],
          "title": "1.10 代码块",
        },
      ],
      "title": "1 Markdown语法教程",
    },
    {
      "anchor": "#2-其他语法",
      "child": [
        {
          "anchor": "#21-html",
          "child": [],
          "title": "2.1 HTML",
        },
        {
          "anchor": "#github-警告",
          "child": [],
          "title": "Github 警告",
        },
      ],
      "title": "2 其他语法",
    },
  ],
}
`;
